### ISSUE TITLE
dev-session harness: GitHub auth (GH_TOKEN env) is lost on workspace suspend/resume — bind a git-credential secret instead

### ISSUE THREAD

**Issue by lenaxia** [lenaxia/LLMSafeSpaces#1087]
## Summary

During a dev session in an LLMSafeSpaces-style workspace pod, **all GitHub authentication evaporated after a workspace suspend → resume**: both `gh` ("not logged in") and `git push` ("could not read Username") failed simultaneously mid-conversation, blocking PR/issue work until the token was manually re-injected.

Root cause (verified by in-pod forensics + source trace, full proof in the comment below):

1. The session harness provisions GitHub auth as a **`GH_TOKEN` environment variable** injected at session creation — never stored anywhere durable. Env injection is **not replayed on resume**.
2. The platform's own durable path for this exact credential exists and works — a bound **`git-credential`-type secret** materializes to `/sandbox-runtime/rt/git-credentials` on every pod boot (`pkg/agentd/secrets/secrets.go:724`, symlinked from `$HOME/.git-credentials` by `cmd/workspace-agentd/init_fs.go:62`, round-trip asserted by `local/test-entrypoint.sh:80-101`) — but **no `git-credential` secret is bound to these dev workspaces**, so the materialize set that correctly restored SSH keys/secrets/config on resume did not include it.
3. Git was additionally coupled to the env var because the session agent had run `gh auth setup-git` (PVC `.gitconfig` → `helper = !gh auth git-credential` → shells out to `gh` → needs `GH_TOKEN`).

`GH_TOKEN` appears nowhere in this repo — the injection is the outer session orchestration. If the session-spawner lives in another repo, transfer this issue there; the platform-side acceptance below is verifiable either way.

## Recommended fix

**Bind a `git-credential` secret (host `github.com`) to dev workspaces at creation** instead of (or in addition to) env-injecting `GH_TOKEN`. Materialization then survives every suspend/resume/OOM-restart by construction — the secret type, token validation, and test coverage already exist; the harness just isn't using them.

Secondary hardening: ship `credential.helper = store --file /home/sandbox/.git-credentials` in the runtime image gitconfig so the materialized file is consumed directly and user-run `gh auth setup-git` can't couple git auth to the env.

## Acceptance criteria

- Suspend → resume a dev workspace with GitHub auth provisioned → `gh auth status` and `git push` both work immediately, with no manual re-injection
- `/sandbox-runtime/rt/git-credentials` exists after resume (no dangling `$HOME/.git-credentials` symlink)
- A regression gate: boot-sequence test asserting the git-credential file materializes from a bound secret on a cold pod (extend the `local/test-entrypoint.sh` pattern or the agentd materialize tests)

**Comment by lenaxia** [comment]
# Proof readout — #1087

Everything below was gathered **from inside the affected workspace pod** during the incident session (2026-08-26 → 2026-08-27, times UTC) plus a source trace against this repo at `main` post-#1072. An agent picking this up should need nothing else. Read-only forensic commands only; no secret values are included (one file was unreadable by design — noted where relevant).

---

## 0. What this sandbox is (environment identification)

The workspace is a production-style LLMSafeSpaces pod, matching the documented volume layout (README-LLM.md "Relay Config Subsystem → Volume layout"):

| Observation | Maps to |
|---|---|
| PID 1 = `/agentd/usr/local/bin/workspace-agentd supervise-opencode` | agentd delivered via image volume (design 0051), uid-separation mode |
| `/workspace` (repo intact), `/home/sandbox`, `/tmp` | PVC subPaths — **persist across suspend/resume** |
| `/sandbox-runtime` (tmpfs) | agent-config.json / secrets / ssh / git-credentials — **wiped on pod death (US-35.7: no plaintext on PVC)** |
| `/sandbox-cfg` | bootstrap emptyDir from the K8s Secret (password, admin-token, free-models.json) |

## 1. Symptom and timeline

| Time (UTC) | Event | Evidence |
|---|---|---|
| Aug 26 ~21:0x | Session start; repo cloned (public, no auth) | session transcript |
| Aug 26 ~21:0x | First `git push` fails: `fatal: could not read Username for 'https://github.com': No such device or address` — **no credential helper configured at all**; note `gh auth status` was ✓ at this point via `GH_TOKEN` | transcript |
| Aug 26 21:13:16 | Pod boot #1: `init_fs` creates `$HOME/.git-credentials -> /sandbox-runtime/rt/git-credentials` | symlink mtime |
| Aug 26 23:01:34 | Session agent runs `gh auth setup-git` → PVC `.gitconfig` gains `helper = !/usr/local/bin/gh auth git-credential` | `.gitconfig` mtime + content (§3.4) |
| Aug 27 06:59:20 | Last commit before suspension | `/workspace/llmsafespaces/.git/HEAD` mtime |
| Aug 27 ~06:0x–06:3x | Last successful `gh` ops (PR #1072 merged, issues closed) — all auth still working | transcript |
| Aug 27 ~15:0x | **Mid-conversation break**: earlier `gh` calls in the same turn succeeded; then `gh issue create` fails with `Alternatively, populate the GH_TOKEN environment variable with a GitHub API authentication token.` → suspend/resume occurred between the calls | transcript |
| Aug 27 15:07:56 | Container restarts | `ps -o lstart= -p 1` (§3.1) |
| Aug 27 15:07 | `/sandbox-runtime` + `/sandbox-cfg` fully re-materialized | dir mtimes (§3.2) |
| Aug 27 15:41 | Investigation underway | `/tmp/opencode` mtime |
| Aug 27 (user refresh) | User re-provisions creds: `gh auth status` ✓ again (`GH_TOKEN`), `git push --dry-run` ✓. **The credential FILE is still absent** — the refresh used env re-injection, the same non-durable path | §3.7 |

## 2. Forensic evidence (verbatim commands and output)

### 2.1 Container restarted at resume

```
$ ps -o pid,lstart,cmd -p 1
      PID  STARTED                      CMD
        1  Thu Aug 27 15:07:56 2026     /agentd/usr/local/bin/workspace-agentd supervise-opencode
```

### 2.2 tmpfs wiped and re-materialized at 15:07 — the materialize set worked

```
$ ls -la /sandbox-runtime/
drwxrwsrwt. 3 root    sandbox  100 Aug 27 15:07 .
-rw-r-----. 1 2000    sandbox 3783 Aug 27 15:07 admin-prompt.md
-rw-r-----. 1 2000    sandbox   10 Aug 27 15:07 allowed-dirs.json
drwxrwx---. 4 sandbox sandbox  140 Aug 27 15:07 rt

$ ls -la /sandbox-runtime/rt/
-rw-------. 1 2000 sandbox   68 Aug 27 15:07 auth.json
drwxr-x---. 2 2000 sandbox   40 Aug 27 15:07 secrets/
-rw-------. 1 2000 sandbox 2100 Aug 27 15:07 secrets.json
drwxr-x---. 2 2000 sandbox   80 Aug 27 15:07 ssh/
-rw-------. 1 2000 sandbox   38 Aug 27 15:07 workspace-config.json

$ ls /sandbox-runtime/rt/ssh/
config  id_ed25519_github        <-- the GitHub SSH key WAS re-materialized (bound secret)

$ ls /sandbox-cfg/                                          <-- bootstrap set, re-created 15:07
admin-token  free-models.json  password
```

Note: `secrets.json` is `0600 uid 2000` (agentd) — unreadable from the session uid by design, so its entry list could not be inspected. The observable outcome (SSH key + secret dirs restored, git-credentials not) is the evidence that the bound-secret set simply does not contain a `git-credential`.

### 2.3 PVC survived — exactly one dangling symlink in $HOME

```
$ ls -la /home/sandbox/ | grep -E "git-credentials|\.ssh|\.secrets"
lrwxrwxrwx. 1 sandbox sandbox 35 Aug 26 21:13 .git-credentials -> /sandbox-runtime/rt/git-credentials
lrwxrwxrwx. 1 sandbox sandbox 27 Aug 26 21:13 .secrets -> /sandbox-runtime/rt/secrets
lrwxrwxrwx. 1 sandbox sandbox 23 Aug 26 21:13 .ssh -> /sandbox-runtime/rt/ssh

$ find /home/sandbox -maxdepth 1 -xtype l
/home/sandbox/.git-credentials          <-- the ONLY dangling one; .secrets and .ssh targets were restored
```

PVC persistence proof: `/tmp/opencode/apt.log` dated Aug 26 23:51 still present; `/workspace/llmsafespaces` repo + branches intact with `.git/HEAD` mtime Aug 27 06:59:20.

### 2.4 Git auth config — the gh shim is the only helper, anywhere

```
$ cat /home/sandbox/.gitconfig
[credential "https://github.com"]
	helper =
	helper = !/usr/local/bin/gh auth git-credential
[credential "https://gist.github.com"]
	helper =
	helper = !/usr/local/bin/gh auth git-credential

$ git config --show-origin --show-scope --get-all credential.helper
(no output — no system/global [credential] helper exists)

$ ls /etc/gitconfig /usr/local/etc/gitconfig /home/sandbox/.config/git/config
ls: cannot access ... : No such file or directory     (all three absent)
```

The shim shells out to `gh`, and `gh` authenticates solely via the `GH_TOKEN` env var (gh config dir `~/.config/gh/hosts.yml` never existed — no stored auth). Hence **gh and git live or die together with that env var**.

### 2.5 The failure pair, verbatim (post-resume)

```
$ gh issue create ...
Alternatively, populate the GH_TOKEN environment variable with a GitHub API authentication token.

$ gh auth status
You are not logged in to any GitHub hosts. To log in, run: gh auth login

$ git push --dry-run origin <branch>
fatal: could not read Username for 'https://github.com': No such device or address

$ ls /sandbox-runtime/rt/git-credentials
ls: cannot access '/sandbox-runtime/rt/git-credentials': No such file or directory
```

### 2.6 Pre-incident auth state (for contrast)

Session start (Aug 26): `gh auth status` → `✓ Logged in to github.com account lenaxia (GH_TOKEN)`, and the env var was present (`gh auth setup-git`-era operations all succeeded for ~9 hours of wall time across 6 merged PRs).

### 2.7 Post-refresh state — the refresh re-proves the mechanism

After the user re-provisioned credentials:

```
$ gh auth status
  ✓ Logged in to github.com account lenaxia (GH_TOKEN)     <-- env var again

$ ls /sandbox-runtime/rt/git-credentials
ls: cannot access ... : No such file or directory           <-- file STILL absent
```

The recovery path is the same fragile env injection; nothing durable changed.

### 2.8 False alarm corrected (recorded per review discipline)

During triage I briefly reported `git config credential.helper` returning `/home/sandbox/.git-credentials`. That was a misparse: the line was `ls ~/.git-credentials` output from an adjacent command in the same shell pipeline, not a config value. Re-verified empty via `--show-origin --show-scope` (§2.4). No config ever referenced the credential file.

## 3. Source trace (repo @ main, post-#1072)

| Location | Role |
|---|---|
| `pkg/secrets/types.go:34` | `SecretTypeGitCredential SecretType = "git-credential"` — first-class secret type |
| `pkg/secrets/types.go:57,71,86` | registered in valid-types set, listed for enumeration, `host` metadata field (optional) |
| `pkg/agentd/secrets/secrets.go:298` | default `GitCredsPath: "/sandbox-runtime/rt/git-credentials"` |
| `pkg/agentd/secrets/secrets.go:724` | materializer writes the credential line on **every boot** from a bound secret: `protocol + "://oauth2:" + token + "@" + host + "\n"`, with URL-safe token validation (conservative allowlist) directly above (lines ~695–723) |
| `pkg/agentd/secrets/secrets.go:554` | `reset()` removes `GitCredsPath` at the top of every materialize — i.e., absent a bound secret, the file is guaranteed absent after each boot, exactly as observed |
| `cmd/workspace-agentd/secrets.go:349` | materialize subcommand path config (`LLMSAFESPACES_GIT_CREDS_PATH` override) |
| `cmd/workspace-agentd/init_fs.go:62` | PVC symlink table: `{pvcRoot}/home/.git-credentials -> {runtimeDir}/rt/git-credentials` — created each boot (observed mtime Aug 26 21:13 on boot #1; symlink survived on PVC, target did not) |
| `local/test-entrypoint.sh:80` | local kind harness binds `{"type":"git-credential","name":"gh","metadata":{"host":"github.com","protocol":"https"},"plaintext":"ghp_abc"}` |
| `local/test-entrypoint.sh:93,101` | asserts `grep -q "ghp_abc" $HOME/.git-credentials` → `GIT_OK` — **the intended round-trip is tested and works when the secret is bound** |
| repo-wide `grep GH_TOKEN` | **zero hits** in Go/sh/YAML (excluding node_modules/tests) → the env injection is the outer session orchestration, not platform code |

Also relevant background: README-LLM.md "Agent-config.json write sequence" — the materialize subcommand runs before agentd on every pod boot from base `secrets.json` + the #443 reload cache; US-35.7 moved credential material to tmpfs deliberately so pod death leaves "dangling symlinks, no plaintext bytes" — which is precisely the observable state here.

## 4. Root-cause chain

1. Suspend deletes the pod; resume creates a new one (PID 1 restart 15:07:56). PVC subPaths persist; `/sandbox-runtime` tmpfs is wiped (by design).
2. The boot materialize runs correctly and restores every **bound** secret: SSH key, secret dirs, config (§2.2). `reset()` (secrets.go:554) deliberately leaves `rt/git-credentials` absent because **no `git-credential` secret is bound to this workspace**.
3. GitHub HTTPS auth for the session was provisioned out-of-band by the session harness as a **`GH_TOKEN` env var at creation time** — durable nowhere, replayed never.
4. `gh` has no stored auth (hosts.yml never existed) → env var loss = `gh` dead.
5. Git's only helper is the `gh auth git-credential` shim (PVC `.gitconfig`, §2.4) → git dead with it.
6. Net effect: one dangling symlink and total GitHub auth loss, mid-conversation, until manual re-injection.

The platform's durable path (bound `git-credential` secret → per-boot materialization → `$HOME/.git-credentials`) exists, is validated, and is covered by the local harness test — the dev-session provisioner simply doesn't use it.

## 5. Reproduction (any agent with a dev workspace)

1. Confirm baseline: `gh auth status` ✓ (GH_TOKEN); `ls -la $HOME/.git-credentials` (dangling); `ls /sandbox-runtime/rt/git-credentials` (absent).
2. Note `git config --show-origin --get-all credential.helper` — empty or gh-shim only; no file-based helper.
3. Ask the operator to suspend + resume the workspace (or wait for a pod restart).
4. Post-resume: `ps -o lstart= -p 1` (new start time), `/sandbox-runtime` fully re-stamped, `gh auth status` → not logged in, `git push` → `could not read Username`.
5. Contrast: `ls /sandbox-runtime/rt/ssh/` — bound secrets DID come back. That asymmetry is the diagnosis.

## 6. Fix direction and acceptance

**Primary (harness):** at dev-workspace creation, create+bind a `git-credential` secret (host `github.com`, protocol `https`, token = the PAT currently env-injected). The materializer then re-creates the file on every boot; `git` works post-resume with zero action. Optionally still inject `GH_TOKEN` for the `gh` CLI — but ALSO ship `credential.helper = store --file /home/sandbox/.git-credentials` in the runtime image gitconfig so git does not depend on `gh`/env at all.

**Acceptance:** (1) after suspend→resume, `gh auth status` and `git push` succeed without manual re-injection; (2) `/sandbox-runtime/rt/git-credentials` exists post-resume (no dangling `$HOME/.git-credentials`); (3) a cold-boot regression gate asserting materialization from a bound secret (extend `local/test-entrypoint.sh` or the agentd materialize tests — the pattern already exists at lines 80–101).

**Note on scope:** if the session-spawner is a separate repo/service, transfer the harness-side work there; the platform-side items (image gitconfig, regression gate) belong here and are independently valuable.

## 7. User-side workaround (meanwhile)

After any resume where creds were re-injected: `gh auth setup-git` if git pushes fail (restores the shim against the fresh env token). Durable self-serve fix from inside the platform: store the PAT as a user secret of type `git-credential` (host `github.com`) and bind it to the workspace — it will then materialize on every boot, which is exactly the cure this issue recommends the harness adopt.
